chronos/achannels.nim

```diff
@@ -0,0 +1,254 @@
+  raise err
+
+proc raiseChannelFailed() {.inline.} =
+  var err = newException(AsyncChannelError, "Channel synchronization failed")
+  raise err
+
+proc open*[Msg](chan: AsyncChannel[Msg]) =
+  ## Open channel ``chan``.
+  let loop = getGlobalDispatcher()
+  chan.acquireLock()
+  inc(chan.refCount)
+  chan.releaseLock()
+
+proc close*[Msg](chan: AsyncChannel[Msg]) =
+  ## Close channel ``chan``.
+  chan.acquireLock()
+  if chan.refCount == 0:
+    if not(isNil(chan.data)):
+      deallocShared(cast[pointer](chan.data))
+    chan.deinitLocks()
+    deallocShared(cast[pointer](chan))
+  else:
+    dec(chan.refCount)
+  chan.releaseLock()
+
+proc `$`*[Msg](chan: AsyncChannel[Msg]): string =
+  ## Dump channel ``chan`` debugging information as string.
+  chan.acquireLock()
+  result = "channel 0x" & toHex(cast[uint](chan)) & " ("
+  result.add("eventNotEmpty = 0x" & toHex(cast[uint](chan.eventNotEmpty)))
+  result.add(", eventNotFull = 0x" & toHex(cast[uint](chan.eventNotFull)))
+  result.add(", rd = " & $chan.rd)
+  result.add(", wr = " & $chan.wr)
+  result.add(", count = " & $chan.count)
+  result.add(", mask = 0x" & toHex(chan.mask))
+  result.add(", data = 0x" & toHex(cast[uint](chan.data)))
+  result.add(", maxItems = " & $chan.maxItems)
+  result.add(", refCount = " & $chan.refCount)
+  result.add(")")
+  chan.releaseLock()
+
+proc rawSend(rchan: RawAsyncChannel, pbytes: pointer, nbytes: int) =
+  var cap = rchan.mask + 1
+  if rchan.count >= cap:
+    if cap == 0: cap = 1
+    var n = cast[ptr UncheckedArray[byte]](allocShared0(cap * 2 * nbytes))
+    var z = 0
+    var i = rchan.rd
+    var c = rchan.count
+
+    while c > 0:
+      dec(c)
+      copyMem(addr(n[z * nbytes]), addr(rchan.data[i * nbytes]), nbytes)
+      i = (i + 1) and rchan.mask
+      inc(z)
+
+    if not isNil(rchan.data):
+      deallocShared(rchan.data)
+
+    rchan.data = n
+    rchan.mask = (cap * 2) - 1
+    rchan.wr = rchan.count
+    rchan.rd = 0
+
+  copyMem(addr(rchan.data[rchan.wr * nbytes]), pbytes, nbytes)
+  inc(rchan.count)
+  rchan.wr = (rchan.wr + 1) and rchan.mask
+
+proc send*[Msg](chan: AsyncChannel[Msg], msg: Msg) {.async.} =
+  ## Send message ``msg`` over channel ``chan``. This procedure will wait if
+  ## internal channel queue is full.
+  chan.acquireLock()
+  try:
+    if chan.refCount == 0:
+      raiseChannelClosed()
+
+    if chan.maxItems > 0:
+      # Wait until count is less then `maxItems`.
+      while chan.count >= chan.maxItems:
+        chan.releaseLock()
+        let res = await chan.eventNotFull.wait(InfiniteDuration)
+        chan.acquireLock()
+        if res == WaitFailed:
+          raiseChannelFailed()
+
+    rawSend(chan, unsafeAddr msg, sizeof(Msg))
+    chan.eventNotEmpty.fire()
+
+  finally:
+    chan.releaseLock()
+
+proc sendSync*[Msg](chan: AsyncChannel[Msg], msg: Msg) =
+  ## Immediately send message ``msg`` over channel ``chan``. This procedure will
+  ## block until internal channel's queue is full.
```

Comment:
Shouldn't it block until the queue is empty because the message was consumed and deleted at the destination?

Reply:
First of all if this function will block when queue is empty, then how we can populate queue? Also delivery confirmation will require adding more synchronization primitives.

Reply:
I misunderstood what `*Sync()` procedures are supposed to do. The comments should be updated to specify they work without an event loop on the caller's side instead of doing unbuffered sending/receiving.